### ISSUE TITLE
TypeScript typing for test.each() doesn't infer heterogenous tuple test case argument types.

### ISSUE THREAD

**Issue by epatey** [oven-sh/bun#5631]
### What version of Bun is running?

1.0.2+37edd5a6e389265738e89265bcbdf2999cb81a49

### What platform is your computer?

Darwin 22.6.0 x86_64 i386

### What steps can reproduce the bug?

```typescript
it.each([
  ["cat", 1],
  ["dog", 2],
])("", (arg0, arg1) => {})
```

### What is the expected behavior?

`arg0` should have an inferred type of `string` and `arg1` should be `number`.

### What do you see instead?

Both `arg0` and `arg1` have the inferred type `string | number`.

### Additional information

If you explicitly declare the test case array as an array of tuples as either of the below samples, the code works as expected.
```typescript
it.each<[string, number]>([
  ["cat", 1],
  ["dog", 2],
])("", (arg0, arg1) => {})
```

```typescript
const cases: [string, number][] = [
  ["cat", 1],
  ["dog", 2],
]
it.each(cases)("", (arg0, arg1) => {})
```

**Comment by benyakirten** [comment]
I encountered a similar problem when trying to interpolate variables. For example:

```ts
test.each([{ a: 1, b: 2, expected: 3 })('returns $expected when adding $a and $b', ({ expected, a, b }) => {
   expected(a+b).toEqual(expected)
}))
```

In other package managers, the line will read `returns 3 when adding 1 and 2`. In Bun, it reads `returns $expected when adding $a and $b` without interpolating the variables.

**Comment by mProjectsCode** [comment]
The typescript bug also happens with `describe.each`.

**Comment by TkDodo** [comment]
my workaround is this:

```
it.each([
    ['a', true],
    ['b', false],
] as const)('label', (...[input, expected]) => {})
```

with this, `input` is inferred as `'a' | 'b'` and `expected` is `boolean`

**Comment by mProjectsCode** [comment]
For reference on how jest types the `each` method.
This is how `@types/jest` types each: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e54876bdf280543e90cd015ca066a4dc5522f868/types/jest/index.d.ts#L542
And this is how `@jest/types` does it: https://github.com/jestjs/jest/blob/0afd8f1c33e0a622727fb696686fe231e4584a70/packages/jest-types/src/Global.ts#L58C10-L58C10

**Comment by mProjectsCode** [comment]
https://github.com/oven-sh/bun/pull/5838

This PR, which was included in the `1.0.3` release fixed the issue but broke typing for an array of objects.

**Comment by nektro** [comment]
does this still reproduce for you? with

```ts
it.each([[{ a: "cat", b: 1 }], [{ a: "dog", b: 2 }]])("", ({ a, b }) => {
  //
});
```

and

```ts
it.each([["cat", 1], ["dog", 2]])("", (arg0, arg1) => {
  //
});
```

using the latest `@type/bun` the type hints in VSCode appear correct for me.
